jetty-websocket/websocket-jetty-api/src/main/java/org/eclipse/jetty/websocket/api/annotations/OnWebSocketMessage.java

```diff
@@ -47,6 +46,16 @@
  * <li><code>public void methodName(InputStream stream)</code></li>
  * <li><code>public void methodName({@link Session} session, InputStream stream)</code></li>
  * </ol>
+ * <u>Partial Message Variations</u>
+ * <p>These are used to receive partial messages without aggregating them into a complete WebSocket message. Instead the a boolean
+ * argument is supplied to indicate whether this is the last segment of data of the message. This can be done most efficiently
+ * for the {@link java.nio.ByteBuffer} type as no copying needs to occur to get the data into this format.</p>
+ * <ol>
+ * <li><code>public void methodName(ByteBuffer payload, boolean last)</code></li>
+ * <li><code>public void methodName(byte payload, boolean last)</code></li>
```

Comment:
Isn't this `byte[]` in the signature?
Frankly, I would not offer a `byte[]` variant.

Also, what do you do with the `ByteBuffer` version if it is retained by the application, but not consumed? I think the signature needs a `Callback` to signal when the `ByteBuffer` can be released, no?

If you choose for no `Callback`, then you have to remove from the javadocs above "as no copying needs to occur to get the data", and specify precisely that applications *must* consume the buffer before returning from the method.

Reply:
> Also, what do you do with the ByteBuffer version if it is retained by the application, but not consumed? I think the signature needs a Callback to signal when the ByteBuffer can be released, no?

The websocket-core API has this feature but for the Jetty and Javax APIs they don't take callbacks, so you only own the `ByteBuffer` only until the method returns. If you wish to retain the `ByteBuffer` you must copy it.

> If you choose for no Callback, then you have to remove from the javadocs above "as no copying needs to occur to get the data"

By this I meant that `ByteBuffer` can be passed into the method directly and doesn't need to be copied by the implementation, but this is not the case for `byte[]` or `String` and there will be some copying involved to get it into that format. I didn't mean that you can retain the `ByteBuffer` indefinitely without copying it.

> and specify precisely that applications must consume the buffer before returning from the method.

They don't have to consume the buffer, if they want to ignore the rest of the message they don't need to consume it. And this is no different to the whole message signatures.

I have removed the `byte[]` option and this part of the Javadoc so it is less confusing.

Reply:
> By this I meant that ByteBuffer can be passed into the method directly and doesn't need to be copied by the implementation, but this is not the case for byte[] or String and there will be some copying involved to get it into that format. I didn't mean that you can retain the ByteBuffer indefinitely without copying it.

This is the javadoc for users -- the implementation details should be left out.
As a user, though, I must know if I have to copy the buffer (assuming interest in the content) before returning from the method.